File: KnnSim.py
from collections import OrderedDict
import itertools
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from pandas import DataFrame
from pandas import Series
import plotnine as gg
import sklearn as sk
from sklearn.neighbors import KNeighborsClassifier

plt.ion()

import SimData


x2_train = SimData.simulate2Group(n = 100,
                                  p = 2,
                                  effect = [1.25] * 2)
knnFit = KNeighborsClassifier(n_neighbors=3)
knnFit.fit(np.array(x2_train['x']), np.array(x2_train['y']))
knnResub = Series(knnFit.predict(x2_train['x']),
                  index = x2_train['y'].index)
np.sum(np.diag(pd.crosstab(knnResub, x2_train['y'])))

x2_test = SimData.simulate2Group(n = 100,
                                 p = 2,
                                 effect = [1.25] * 2)
knnTest = Series(knnFit.predict(x2_test['x']),
                 index = x2_test['y'].index)
np.sum(np.diag(pd.crosstab(knnTest, x2_test['y'])))


def expandGrid(od):
    cartProd = list(itertools.product(*od.values()))
    return DataFrame(cartProd, columns=od.keys())

parVals = OrderedDict()
parVals['n'] = [100]
parVals['p'] = [2, 5, 10, 25, 100, 500]
parVals['k'] = [3, 5, 10, 25]
parGrid = expandGrid(parVals)
parGrid['effect'] = 2.5
parGrid['effect'] = parGrid['effect'] / np.sqrt(parGrid['p'])


def knnSimulate(param):
    trainSet = SimData.simulate2Group(
        n = int(param['n']),
        p = int(param['p']),
        effect = [param['effect']] * int(param['p'])
    )
    knnFit = KNeighborsClassifier(n_neighbors=int(param['k']))
    knnFit.fit(np.array(trainSet['x']), np.array(trainSet['y']))
    testSet = SimData.simulate2Group(
        n = int(param['n']),
        p = int(param['p']),
        effect = [param['effect']] * int(param['p'])
    )
    out = OrderedDict()
    out['p'] = int(param['p'])
    out['k'] = int(param['k'])
    out['train'] = trainSet
    out['test'] = testSet
    out['resubPreds'] = knnFit.predict(trainSet['x'])
    out['resubProbs'] = knnFit.predict_proba(trainSet['x'])
    out['testPreds'] = knnFit.predict(testSet['x'])
    out['testProbs'] = knnFit.predict_proba(testSet['x'])
    out['resubTable'] = pd.crosstab(
        Series(out['resubPreds'], index=trainSet['y'].index),
        trainSet['y']
    )
    out['resubAccuracy'] = (np.sum(np.diag(out['resubTable'])) /
                            (1.0 * np.sum(np.sum(out['resubTable']))))
    out['testTable'] = pd.crosstab(
        Series(out['testPreds'], index=testSet['y'].index),
        testSet['y']
    )
    out['testAccuracy'] = (np.sum(np.diag(out['testTable'])) /
                           (1.0 * np.sum(np.sum(out['testTable']))))
    return out


knnResults = [knnSimulate(parGrid.iloc[i])
              for i in range(parGrid.shape[0])]


repeatedKnnResults = []
for r in range(10):
    repeatedKnnResults.extend(knnSimulate(parGrid.iloc[i])
                              for i in range(parGrid.shape[0]))

knnResultsSimplified = DataFrame([(x['p'],
                                   x['k'],
                                   x['resubAccuracy'],
                                   x['testAccuracy'])
                                  for x in repeatedKnnResults],
                                 columns = ['p',
                                            'k',
                                            'resubAccuracy',
                                            'testAccuracy'])

ggdata = pd.concat(
    [DataFrame({'p' : knnResultsSimplified.p,
                'k' : knnResultsSimplified.k.apply(int),
                'type' : 'resub',
                'Accuracy' : knnResultsSimplified.resubAccuracy}),
     DataFrame({'p' : knnResultsSimplified.p,
                'k' : knnResultsSimplified.k.apply(int),
                'type' : 'test',
                'Accuracy' : knnResultsSimplified.testAccuracy})],
    axis = 0
)

plt.close()
ggo = gg.ggplot(ggdata, gg.aes(x='p', y='Accuracy',
                               color='type', group='type', linetype='type'))
ggo += gg.facet_wrap('~ k')
ggo += gg.scale_x_log10()
ggo += gg.geom_point(alpha=0.6)
ggo += gg.stat_smooth()
ggo += gg.theme_bw()
print(ggo)

File: KnnSim.R
library(class)
library(ggplot2)

source("SimData.R")


x2_train = simulate2Group(n=100, p=2, effect=rep(1.25, 2))
knnResub = knn(
    train = x2_train$x,
    test = x2_train$x,
    cl = x2_train$y,
    k = 3
)
sum(diag(table(knnResub, x2_train$y)))
x2_test = simulate2Group(n=100, p=2, effect=rep(1.25, 2))
knnTest = knn(
    train = x2_train$x,
    test = x2_test$x,
    cl = x2_train$y,
    k = 3
)
sum(diag(table(knnTest, x2_test$y)))


parGrid = expand.grid(
    n = 100,
    p = c(2, 5, 10, 25, 100, 500),
    k = c(3, 5, 10, 25)
)
parGrid$effect = rep(2.5 / sqrt(parGrid$p))
rownames(parGrid) = paste0("p", parGrid$p, "_k", parGrid$k)

knnSimulate = function(param) {
    param = as.list(param)
    trainSet = simulate2Group(n=param$n, p=param$p,
            effect=rep(param$effect, param$p))
    testSet = simulate2Group(n=param$n, p=param$p,
            effect=rep(param$effect, param$p))
    out = list(
        p = param$p,
        k = param$k,
        train = trainSet,
        test = testSet,
        resubPreds = knn(trainSet$x, trainSet$x, trainSet$y,
                k=param$k),
        resubProbs = knn(trainSet$x, trainSet$x, trainSet$y,
                k=param$k, prob=TRUE),
        testPreds = knn(trainSet$x, testSet$x, trainSet$y,
                k=param$k),
        testProbs = knn(trainSet$x, testSet$x, trainSet$y,
                k=param$k, prob=TRUE)
    )
    out$resubTable = table(
        Predicted = out$resubPreds,
        Actual = trainSet$y
    )
    out$resubAccuracy = sum(diag(out$resubTable)) /
            sum(out$resubTable)
    out$testTable = table(
        Predicted = out$testPreds,
        Actual = testSet$y
    )
    out$testAccuracy = sum(diag(out$testTable)) /
            sum(out$testTable)
    return(out)
}


set.seed(123)
knnResults = apply(X=parGrid, MARGIN=1, FUN=knnSimulate)


set.seed(123)
repeatedKnnResults = lapply(X=1:10, FUN=function(...) {
    apply(X=parGrid, MARGIN=1, FUN=knnSimulate)
})
repeatedKnnResults = do.call(c, args=repeatedKnnResults)
knnResultsSimplified = data.frame(do.call(rbind, args=lapply(
    X = repeatedKnnResults,
    FUN = function(x) {
        outColnames = c("p", "k", "resubAccuracy", "testAccuracy")
        out = x[outColnames]
        return(structure(as.numeric(out), names=outColnames))
    }
)))


ggdata = rbind(
    data.frame(
        p = knnResultsSimplified$p,
        k = paste0("k=", knnResultsSimplified$k),
        type = "resub",
        Accuracy = knnResultsSimplified$resubAccuracy
    ),
    data.frame(
        p = knnResultsSimplified$p,
        k = paste0("k=", knnResultsSimplified$k),
        type = "test",
        Accuracy = knnResultsSimplified$testAccuracy
    )
)
ggdata$k = factor(as.character(ggdata$k),
        levels=c("k=3", "k=5", "k=10", "k=25"))

ggobj = ggplot(
    data = ggdata,
    mapping = aes(x=p, y=Accuracy,
            color=type, group=type, linetype=type)
) + theme_bw()
ggobj = ggobj + scale_x_log10()
ggobj = ggobj + geom_point(alpha=0.6)
ggobj = ggobj + stat_smooth(method.args=list(degree=1))
ggobj = ggobj + facet_wrap(~k)
## pdf("KnnSim.pdf", h=5, w=5*1.175)
print(ggobj)
## garbage = dev.off()

